ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveReduceExpressionsWithStatsRule.java

```diff
@@ -272,7 +274,11 @@ public RexNode visitCall(RexCall call) {
       // If we did not reduce, check the children nodes
       RexNode node = super.visitCall(call);
       if (node != call) {
-        node = RexUtil.simplify(rexBuilder, node);
+        // FIXME if this rule will make some changes; then it will invoke simplify on all subtrees during exiting the recursion
```

Comment:
Is this to fix in this patch or in a follow-up? Can you create a JIRA if that is the case?

Reply:
It was there before; and I can't really do anything right now with it...leaving a comment might be usefull if it lights up on someones profiler

To fix this we would need some way to keep track which nodes were simplified already...that should be fixed in Calcite;and it might not even possible (because the set of predicates might be different )

We might consider moving the simplification outside of the recursive phase - that will certainly reduce the re-visit pressure; but could result in reduced simplification in some cases (however considering how many places we are doing this, that's unlikely) - it could be explored in a separate patch.